crates/turbo-tasks/src/join_iter_ext.rs

```diff
@@ -96,17 +92,19 @@ where
                 match result {
                     Ok(children) => {
                         for item in children {
-                            let (index, new) = this.set.insert_full(item);
-                            if new {
-                                this.futures
-                                    .push_back((this.get_children)(this.set[index].clone()));
+                            match (this.filter_flat_map)(&item) {
+                                Some(future) => {
+                                    this.futures.push(future);
+                                }
+                                None => {}
                             }
+                            this.output.push(item);
                         }
                     }
                     Err(err) => return std::task::Poll::Ready(Err(err)),
                 }
             } else {
-                return std::task::Poll::Ready(Ok(take(&mut this.set)));
+                return std::task::Poll::Ready(Ok(std::mem::take(&mut this.output)));
```

Comment:
... and topo sort it here?

Reply:
Yeah, that sounds like a good idea. So you have to opt-in into `Sort::Unordered` and it's visible from usage that it's non-deterministic.

`try_flat_map_recursive_join(Unordered, |...| { ... })`
or
`try_flat_map_recursive_join(Topologically, |...| { ... })`

Reply:
Yes. Although I'd probably go for `try_flat_map_recursive_join::<Unordered>::(...)` if that's something we can make generic enough.